interpreters/wamr/wamr_custom_init.c

```diff
@@ -0,0 +1,657 @@
+    }
+}
+
+#ifndef GLUE_FUNCTION_sendmsg
+#define GLUE_FUNCTION_sendmsg
+uintptr_t glue_sendmsg(wasm_exec_env_t env, uintptr_t parm1,
+                       uintptr_t parm2, uintptr_t parm3)
+{
+  wasm_module_inst_t module_inst = get_module_inst(env);
+  FAR struct msghdr *hdr =
+    (FAR struct msghdr *)(uintptr_t)parm2;
+
+  glue_msghdr_begin(module_inst, hdr);
+  uintptr_t ret = sendmsg((int)parm1,
+                          (FAR struct msghdr *)(parm2),
+                          (int)parm3);
+  glue_msghdr_end(module_inst, hdr);
+
+  return ret;
+}
+
+#endif /* GLUE_FUNCTION_sendmsg */
+
+#ifndef GLUE_FUNCTION_recvmsg
+#define GLUE_FUNCTION_recvmsg
+uintptr_t glue_recvmsg(wasm_exec_env_t env, uintptr_t parm1,
+                       uintptr_t parm2, uintptr_t parm3)
+{
+  wasm_module_inst_t module_inst = get_module_inst(env);
+  FAR struct msghdr *hdr =
+    (FAR struct msghdr *)((uintptr_t)parm2);
+
+  glue_msghdr_begin(module_inst, hdr);
+  uintptr_t ret = recvmsg((int)parm1,
+                          (FAR struct msghdr *)(parm2),
+                          (int)parm3);
+  glue_msghdr_end(module_inst, hdr);
+
+  return ret;
+}
+
+#endif /* GLUE_FUNCTION_recvmsg */
+
+#ifndef GLUE_FUNCTION_strsep
+#define GLUE_FUNCTION_strsep
+uintptr_t glue_strsep(wasm_exec_env_t env, uintptr_t parm1, uintptr_t parm2)
+{
+  wasm_module_inst_t module_inst = get_module_inst(env);
+  FAR char **stringp = parm1;
+
+  if (*stringp != NULL)
+    {
+      *stringp = addr_app_to_native(*stringp);
+    }
+
+  return addr_native_to_app((uintptr_t)strsep(
+    (FAR char **)addr_app_to_native(parm1),
+    (FAR const char *)addr_app_to_native(parm2)));
+}
+
+#endif /* GLUE_FUNCTION_strsep */
+
+#ifndef GLUE_FUNCTION_scandir
+#define GLUE_FUNCTION_scandir
+uintptr_t glue_scandir(wasm_exec_env_t env, uintptr_t parm1, uintptr_t parm2,
+                       uintptr_t parm3, uintptr_t parm4)
+{
+  wasm_module_inst_t module_inst = get_module_inst(env);
+  int ret = 0;
+  int i = 0;
+
+  ret = scandir((FAR const char *)parm1,
+                 (FAR struct dirent ***)parm2,
+                 (FAR void *)parm3, alphasort);
+
+  for (i = 0; i < ret; i++)
+    {
+      (*(uintptr_t **)parm2)[i] = addr_native_to_app((*(uintptr_t **)parm2)[i]);
+    }
+
+  *(uintptr_t *)parm2 = addr_native_to_app(*(uintptr_t *)parm2);
+
+  return ret;
+}
+
+#endif /* GLUE_FUNCTION_scandir */
+
+#ifndef GLUE_FUNCTION_daemon
+#define GLUE_FUNCTION_daemon
+uintptr_t glue_daemon(wasm_exec_env_t env, uintptr_t parm1, uintptr_t parm2)
+{
+  return 0;
+}
+
+#endif /* GLUE_FUNCTION_daemon */
+
+#ifndef GLUE_FUNCTION_posix_memalign
+#define GLUE_FUNCTION_posix_memalign
+uintptr_t glue_posix_memalign(wasm_exec_env_t env, uintptr_t parm1, uintptr_t parm2, uintptr_t parm3)
+{
+  uintptr_t rawptr;
+  wasm_module_inst_t module_inst = get_module_inst(env);
+
+  /* Extra size for align */
+
+  parm3 += parm2;
+  int ret = posix_memalign((FAR void **)parm1, (size_t)parm2, (size_t)parm3);
```

Comment:
doesn't this return an aligned pointer?

Reply:
posix_memalign return an aligned pointer, but it may not an aligned value after convert to be offset of linear memory.